### PR TITLE
feat: add softDelete

### PR DESCRIPTION
**Pull request by JoaoFerrazfs**
## Description

This PR is part of a series of PRs aimed at adding the softDelete functionality to Mongolid.

---

The following features are being added in this implementation:

- SoftDelete: The `delete` method, when needed, will perform softDelete.
- Restoration of deleted objects, achieved through the `restore` method.
- Filtering that prevents results deleted with softDelete from being returned in searches.
- The `withTrashed` method that returns all results, including those deleted with softDelete.
- The `forceDelete` method that forcefully deletes, without considering softDelete.

---

### CODE REVIEW

**Review by jhonatanveras — DISMISSED** [review]

File: src/DataMapper/DataMapper.php [hunk 1]
```diff
@@ -665,4 +616,12 @@ function ($value) {
...
         return $filtered;
     }
+
+    private function executeSoftDelete(ModelInterface $entity, $options): bool
+    {
+        $deletedAtCoullum = QueryBuilder::getDeletedAtColumn($entity);
+        $entity->$deletedAtCoullum = new UTCDateTime(new DateTime('now'));
+
+        return $this->update($entity, $options);
+    }
```
Comment (GetulioMeirelles) [inline src/DataMapper/DataMapper.php]: Talvez faça mais sentido deixar esse método dentro do trait.

**Review by GetulioMeirelles — DISMISSED** [review]

File: src/DataMapper/DataMapper.php (hunk 1 above)
Comment (JoaoFerrazfs) [reply src/DataMapper/DataMapper.php]: Fiz essa alteração aqui, mas adicionei no PR parte 2 [PR](https://github.com/leroy-merlin-br/mongolid/pull/214).
Como la ja tem os dois tipos de models eu aproveitei pra deixar a traint la e também modifiquei as duas para ter a compatibilidade

File: src/Util/QueryBuilder.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace Mongolid\Util;
+
+use MongoDB\BSON\ObjectId;
+use Mongolid\Model\ModelInterface;
+
+class QueryBuilder
+{
+    public static function resolveQuery(int|array|string $query, ModelInterface $model): array
```
Comment (orlandocavassani) [inline src/Util/QueryBuilder.php]:
```suggestion
    public static function resolveQuery(mix $query, ModelInterface $model): array
```
I think we can pass more types than specified here, like `MongoDB\BSON\ObjectId`. What do you think to change it to _mixed_?

File: src/Util/QueryBuilder.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            ? $model::DELETED_AT
+            : 'deleted_at';
+    }
+
+    /**
+     * Transforms a value that is not an array into an MongoDB query (array).
+     * This method will take care of converting a single value into a query for
+     * an _id, including when a objectId is passed as a string.
+     */
+    public static function prepareValueQuery(int|array|string $value): array
```
Comment (orlandocavassani) [inline src/Util/QueryBuilder.php]:
```suggestion
    public static function prepareValueQuery(mixed $value): array
```

File: src/Model/SoftDeletesTrait.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+namespace Mongolid\Model;
+
+use DateTime;
+use MongoDB\BSON\UTCDateTime;
+use Mongolid\Cursor\CursorInterface;
+use Mongolid\Util\QueryBuilder;
+
+trait SoftDeletesTrait
+{
+    public bool $enabledSoftDeletes = true;
```
Comment (orlandocavassani) [inline src/Model/SoftDeletesTrait.php]:
```suggestion
    protected bool $isSoftDeleteEnabled = true;
```
Maybe we can change this attribute to protected. What do you think?

File: src/Model/SoftDeletesTrait.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    public bool $forceDelete = false;
```
Comment (orlandocavassani) [inline src/Model/SoftDeletesTrait.php]: I think you don't need `forceDelete` attribute. The `enabledSoftDelete` should be enough.

File: src/Model/SoftDeletesTrait.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        $collumn = self::getDeletedAtColumn();
+
+        if (!$this->isTrashed()) {
+            return false;
+        }
+
+        unset($this->{$collumn});
+
+        return $this->execute('save');
```
Comment (orlandocavassani) [inline src/Model/SoftDeletesTrait.php]:
```suggestion
        return $this->update();
```
We could use update method here?

File: src/Model/SoftDeletesTrait.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        $this->forceDelete = true;
+
+        return $this->execute('delete');
+    }
+
+    public function executeSoftDelete(): bool
+    {
+        $deletedAtCoullum = QueryBuilder::getDeletedAtColumn($this);
+        $this->$deletedAtCoullum = new UTCDateTime(new DateTime('now'));
```
Comment (orlandocavassani) [inline src/Model/SoftDeletesTrait.php]:
```suggestion
        $this->$deletedAtCoullum = new UTCDateTime();
```
:nail_care:

File: src/Model/SoftDeletesTrait.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+        return $this->update();
+    }
+
+    public static function withTrashed(
+        array|string|int $query = [],
```
Comment (orlandocavassani) [inline src/Model/SoftDeletesTrait.php]:
```suggestion
        mixed $query = [],
```

File: src/Model/SoftDeletesTrait.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    private static function getDeletedAtColumn(): string
+    {
+        return defined(
+            static::class . '::DELETED_AT'
+        )
+            ? static::DELETED_AT
+            : 'deleted_at';
+    }
```
Comment (orlandocavassani) [inline src/Model/SoftDeletesTrait.php]: This method already exists at `src/Util/QueryBuilder.php`

File: src/Model/SoftDeletesTrait.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    public function forceDelete(): bool
+    {
+        $this->forceDelete = true;
+
+        return $this->execute('delete');
+    }
+
+    public function executeSoftDelete(): bool
+    {
+        $deletedAtCoullum = QueryBuilder::getDeletedAtColumn($this);
```
Comment (orlandocavassani) [inline src/Model/SoftDeletesTrait.php]:
```suggestion
        $deletedAtColumn = QueryBuilder::getDeletedAtColumn($this);
```

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+use Mongolid\Tests\Stubs\Legacy\Product;
+
+final class PersisteLegacyModelWithSoftDeleteTest extends IntegrationTestCase
+{
+    private ObjectId $_id;
+
+    public function testShouldFindNotDeletedProduct(): void
+    {
+        // Set
+        $product = $this->persiteProduct();
```
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php]:
```suggestion
        $product = $this->persistProduct();
```

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+use MongoDB\BSON\UTCDateTime;
+use Mongolid\Model\ModelInterface;
+use Mongolid\Tests\Stubs\Legacy\ProductWithSoftDelete;
+use Mongolid\Tests\Stubs\Legacy\Product;
+
+final class PersisteLegacyModelWithSoftDeleteTest extends IntegrationTestCase
+{
+    private ObjectId $_id;
+
+    public function testShouldFindNotDeletedProduct(): void
```
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php]:
```suggestion
    public function testShouldFindUndeletedProduct(): void
```

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php:1]: The word `Persiste` in file name is incorrect.

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        // Actions
+        $actualWhereResult = ProductWithSoftDelete::where()->first();
+        $actualFirstResult = ProductWithSoftDelete::first($this->_id);
+
+        // Assertions
+        $this->assertEquals($product, $actualWhereResult);
+        $this->assertEquals($product, $actualFirstResult);
+    }
+
+    public function testCannotFindDeletedProduct(): void
```
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php]:
```suggestion
    public function testShouldNotFindDeletedProduct(): void
```

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        // Assertions
+        $this->assertSame(2, $result->count());
+        $this->assertInstanceOf(
+            UTCDateTime::class,
+            $resultArray[0]['deleted_at']
+        );
+        $this->assertNull($resultArray[1]['deleted_at'] ?? null);
+    }
+
+    public function testRestoreDeletedProduct(): void
```
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php]:
```suggestion
    public function testShouldRestoreDeletedProduct(): void
```

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+        // Actions
+        $isRestored = $product->restore();
+        $result = ProductWithSoftDelete::first($this->_id);
+
+        // Assertions
+        $this->assertTrue($isRestored);
+        $this->assertEquals($product, $result);
+    }
+
+    public function testCannotRestoreAlreadyRestoredProduct(): void
```
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php]:
```suggestion
    public function testShouldNotRestoreAlreadyRestoredProduct(): void
```

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+        // Actions
+        $isRestored = $product->restore();
+        $result = ProductWithSoftDelete::first($this->_id);
+
+        // Assertions
+        $this->assertFalse($isRestored);
+        $this->assertEquals($product, $result);
+    }
+
+    public function testExecuteSoftDeleteOnProduct(): void
```
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php]:
```suggestion
    public function testShouldExecuteSoftDeleteOnProduct(): void
```

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+        $this->assertTrue($isDeleted);
+        $this->assertNull($result);
+        $this->assertEquals($product, $deletedProduct);
+        $this->assertInstanceOf(
+            UTCDateTime::class,
+            $deletedProduct->deleted_at
+        );
+    }
+
+    public function testExecuteForceDeleteOnProduct(): void
```
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php]:
```suggestion
    public function testShouldExecuteForceDeleteOnProduct(): void
```

File: tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+         $isDeleted = $product->forceDelete();
+         $result = ProductWithSoftDelete::withTrashed();
+
+        // Assertions
+        $this->assertTrue($isDeleted);
+        $this->assertSame(1, $result->count());
+        $this->assertEquals($product2, $result->first());
+    }
+
+    public function testCannotExecuteSoftDeleteOnProduct(): void
```
Comment (orlandocavassani) [inline tests/Integration/PersisteLegacyModelWithSoftDeleteTest.php]:
```suggestion
    public function testShouldNotExecuteSoftDeleteOnProduct(): void
```

File: tests/Integration/PersisteModelWithSoftDeleteTest.php
Comment (orlandocavassani) [inline tests/Integration/PersisteModelWithSoftDeleteTest.php:1]: Try to apply the same suggestions at legacy test.

File: tests/Unit/Model/SoftDeletesTraitTest.php [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+        $product = new ProductWithSoftDelete();
+
+        // Actions
+        $actual = $product->restore();
+
+        // Assertions
+        $this->assertFalse($actual);
+    }
+
+    public function testShouldforceDeleteProduct(): void
```
Comment (orlandocavassani) [inline tests/Unit/Model/SoftDeletesTraitTest.php]:
```suggestion
    public function testShouldForceDeleteOnProduct(): void
```

File: src/Util/QueryBuilder.php (hunk 2 above)
Comment (JoaoFerrazfs) [reply src/Util/QueryBuilder.php]: Change made

File: src/Util/QueryBuilder.php (hunk 3 above)
Comment (JoaoFerrazfs) [reply src/Util/QueryBuilder.php]: Change made

File: src/Model/SoftDeletesTrait.php (hunk 5 above)
Comment (JoaoFerrazfs) [reply src/Model/SoftDeletesTrait.php]: You are right

File: src/Model/SoftDeletesTrait.php (hunk 6 above)
Comment (JoaoFerrazfs) [reply src/Model/SoftDeletesTrait.php]: Yes I tested here and the suggestion worked

File: src/Model/SoftDeletesTrait.php (hunk 9 above)
Comment (JoaoFerrazfs) [reply src/Model/SoftDeletesTrait.php]: Removed the duplicated method

File: tests/Integration/PersistLegacyModelWithSoftDeleteTest.php [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+         $isDeleted = $product->delete();
+         $result = ProductWithSoftDelete::first('5bcb310783a7fcdf1bf1a672');
+
+        // Assertions
+        $this->assertTrue($isDeleted);
+        $this->assertNull($result);
+        $this->assertNull($result->deleted_at ?? null);
+    }
+
+    private function persitProductWithSoftDeleteTrait(
```
Comment (orlandocavassani) [inline tests/Integration/PersistLegacyModelWithSoftDeleteTest.php]:
```suggestion
    private function persistProductWithSoftDeleteTrait(
```

File: tests/Integration/PersistLegacyModelWithSoftDeleteTest.php [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+        $product->_id = new ObjectId($id);
+        $product->short_name = 'Furadeira de Impacto Bosch com Chave de Mandril ';
+        $product->name = 'Furadeira de Impacto Bosch com Chave de Mandril e Acessórios 550W 1/2 GSB 550 RE 127V (110V)';
+
+        $product->save();
+
+        return $product;
+    }
+
+    private function persitProduct(
```
Comment (orlandocavassani) [inline tests/Integration/PersistLegacyModelWithSoftDeleteTest.php]:
```suggestion
    private function persistProduct(
```

File: tests/Integration/PersistModelWithSoftDeleteTest.php [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+        $product = new ProductWithSoftDelete();
+        $product->_id = new ObjectId($id);
+        $product->short_name = 'Furadeira de Impacto Bosch com Chave de Mandril ';
+        $product->name = 'Furadeira de Impacto Bosch com Chave de Mandril e Acessórios 550W 1/2 GSB 550 RE 127V (110V)';
+
+        $product->save();
+
+        return $product;
+    }
+    private function persitProduct(
```
Comment (orlandocavassani) [inline tests/Integration/PersistModelWithSoftDeleteTest.php]:
```suggestion
    private function persistProduct(
```

File: tests/Integration/PersistModelWithSoftDeleteTest.php [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->_id = new ObjectId('5bcb310783a7fcdf1bf1a672');
+    }
+
+    private function persitProductWithSoftDeleteTrait(
```
Comment (orlandocavassani) [inline tests/Integration/PersistModelWithSoftDeleteTest.php]:
```suggestion
    private function persistProductWithSoftDeleteTrait(
```

File: tests/Unit/DataMapper/DataMapperTest.php [hunk 24]
```diff
@@ -604,6 +601,7 @@ public function testShouldDelete($entity, $writeConcern, $shouldFireEventAfter,
         $options = ['writeConcern' => new WriteConcern($writeConcern)];
 
         $entity->_id = null;
+        $entity->forceDelete = true;
```
Comment (orlandocavassani) [inline tests/Unit/DataMapper/DataMapperTest.php]:
```suggestion
```

**Review by orlandocavassani — DISMISSED** [review]

**Review by orlandocavassani — APPROVED** [review]

**Review by GetulioMeirelles — APPROVED** [review]

**Review by jhonatanveras — APPROVED** [review]

**Review by GetulioMR — APPROVED** [review]

**Pull request merged**